resharper/src/resharper-unity/Daemon/Stages/Analysis/UnityNullCoalescingProblemAnalyzer.cs

```diff
@@ -0,0 +1,42 @@
+﻿using JetBrains.Annotations;
+using JetBrains.ReSharper.Feature.Services.Daemon;
+using JetBrains.ReSharper.Plugins.Unity.Daemon.Stages.Dispatcher;
+using JetBrains.ReSharper.Plugins.Unity.Daemon.Stages.Highlightings;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+using JetBrains.ReSharper.Psi.Resolve;
+using JetBrains.ReSharper.Psi.Util;
+
+namespace JetBrains.ReSharper.Plugins.Unity.Daemon.Stages.Analysis
+{
+    [ElementProblemAnalyzer(typeof(INullCoalescingExpression), HighlightingTypes = new[] { typeof(UnityNullCoalescingWarning) })]
+    public class UnityNullCoalescingProblemAnalyzer : UnityElementProblemAnalyzer<INullCoalescingExpression>
+    {
+        public UnityNullCoalescingProblemAnalyzer([NotNull] UnityApi unityApi)
+            : base(unityApi)
+        {
+        }
+
+        protected override void Analyze(INullCoalescingExpression expression, ElementProblemAnalyzerData data, IHighlightingConsumer consumer)
+        {
+            if (expression.LeftOperand == null || expression.RightOperand == null)
+                return;
+            if (!(expression.LeftOperand is IReferenceExpression leftOperand) || !IsDescendantOfUnityObject(leftOperand))
+                return;
+
+            consumer.AddHighlighting(new UnityNullCoalescingWarning(expression));
+        }
+
+        private static bool IsDescendantOfUnityObject([CanBeNull]IReferenceExpression expression)
```

Comment:
I was trying to use `Api.IsUnityType` here but didn't seem to get it to work and ended up implementing my own version. Specifically the check `if (type.Module is IProjectPsiModule projectPsiModule)` fails.
Also the same method exists in `UnityNullPropagationProblemAnalyzer.cs` so if we keep it it should move to a central location.

Reply:
This is because the `UnityApi` class is designed to contain the messages/event functions that a class derived from a Unity type can implement, rather than describe all of the Unity API. I think it would be worth renaming that method to be something like `IsDerivedFromUnityType`, or `IsUserUnityType`.

What you've got looks ok, but you could also do [something like this](https://github.com/JetBrains/resharper-unity/blob/4c3fd20248bf793fc0290e8b91502ca99273f73c/resharper/src/resharper-unity/Daemon/Stages/Analysis/IncorrectObjectInstantiationProblemAnalyzer.cs#L29-L33), creating an instance of `IType` for the known name, and using that to pass into a call to `Type.IsSubtypeOf`.

But don't worry about fixing this up now, I want to get this merged. I might have a little refactor of `UnityApi` and add this in there, or something.

Reply:
Yeah, that example is what I was looking for. It's quite hard to understand the relations between all the different types. I still don't really get it. Anyway I gave it a quick test and it works. 

```cs
        protected override void Analyze(INullCoalescingExpression expression, ElementProblemAnalyzerData data, IHighlightingConsumer consumer)
        {
            if (!(expression.LeftOperand is IReferenceExpression leftOperand) || expression.RightOperand == null)
                return;
            var resolve = leftOperand.Reference.Resolve();
            if (resolve.ResolveErrorType != ResolveErrorType.OK)
                return;
            var unityObjectType = TypeFactory.CreateTypeByCLRName(KnownTypes.Object, expression.GetPsiModule());
            if (!leftOperand.Type().IsSubtypeOf(unityObjectType))
                return;

            consumer.AddHighlighting(new UnityNullCoalescingWarning(expression));
        }
```

Unless you want to refactor it later anyway I could commit this to remove some of the redundancy if you want.

Reply:
If you've got the commit ready, fire away! 👍

Reply:
Yes, the type system interfaces can be confusing. There are two hierarchies in play (with `ITreeNode` mixed in there, too), and they both serve different purposes, and the both have a LOT of derived interfaces.

The `IDeclaredElement` interface provides a language agnostic base for "things that are declared". This can be types, like classes (`ITypeElement`), or variables (`ILocalVariable`), or namespaces (`INamespace`) or file system paths (`IPathDeclaredElement`) or even colours (`IColorDeclaredElement`). But these derived interfaces represent the declarations, and aren't good at representing usages (e.g. `MyClass[]` or `List<string>`).

For that, we have another interface hierarchy, which is a full semantic model. This hierarchy is language specific, and the CLR implementation is `IType` and its base class `IExpressionType`. This can represent all type usages, including arrays (`IArrayType`) and pointers (`IPointerType`). The JavaScript type system has a different, but similar hierarchy, rooted at `IJavaScriptType`.

You can bounce between the two hierarchies - the `IType` hierarchy allows getting back to declared elements (if there are any) via `IDeclaredType.GetTypeElement()`, while the `IDeclaredElement` hierarchy can get to the semantic usage model via `ITypeOwner.Type`.

I usually have to look at base and deriving classes to remember which hierarchy I'm in :)

There's some more detail in the devGuide about [declared elements](https://www.jetbrains.com/help/resharper/sdk/PSI/DeclaredElements.html) and [type systems](https://www.jetbrains.com/help/resharper/sdk/PSI/TypeSystems2.html).